File: Credit-Processing/clientdata.cpp
// Fig. 14.9: ClientData.cpp
// Class Client Data stores customer's credit information.
#include <string>
#include <stdexcept>
#include "clientdata.h"

// default ClientData constructor
ClientData::ClientData(double balanceValue, const std::string &lastName, const std::string &firstName ,int accountNumberValue)
    : accountNumber(accountNumberValue), balance(balanceValue)
{
    setLastName(lastName);
    setFirstName(firstName);
}

// get account-number value
int ClientData::getAccountNumber() const { return accountNumber; }

// set account-number value
void ClientData::setAccountNumber(int accountNumberValue)
{
    if (accountNumberValue < 0.0)
        throw std::invalid_argument("Account Number must be >= 0");
    accountNumber = {accountNumberValue}; // should validate
}

// get last-name value
std::string ClientData::getLastName() const { return lastName; }

// set last-name value
void ClientData::setLastName(const std::string &lastNameString)
{
    // copy at most 15 characters from string to lastName
    size_t length(lastNameString.size());
    length = (length < 15 ? length : 14);
    lastNameString.copy(lastName, length);
    lastName[length] = '\0'; // append null character to lastName
}

// get first-name value
std::string ClientData::getFirstName() const { return firstName; }

// set first-name value
void ClientData::setFirstName(const std::string &firstNameString)
{
    // copy at most 10 characters from string to firstName
    size_t length{firstNameString.size()};
    length = (length < 10 ? length : 9);
    firstNameString.copy(firstName, length);
    firstName[length] = '\0'; // append null character to firstName
}

// get balance value
double ClientData::getBalance() const { return balance; }

// set balance value
void ClientData::setBalance(double balanceValue) { balance = balanceValue; }

File: Models/ClientData.cs
using System;

namespace comp1210.Models;
public class ClientData
{
  // Constants
  public static int RECORD_COUNT = 15;
  // Private fields
  private int accountNumber;
  private string lastName = string.Empty;
  private string firstName = string.Empty;
  private double balance;

  // Default constructor with optional parameters
  public ClientData(double balance = 0, string lastName = "", string firstName = "", int accountNumber = 0)
  {
    AccountNumber = accountNumber;
    LastName = lastName;
    FirstName = firstName;
    Balance = balance;
  }

  // AccountNumber property
  public int AccountNumber
  {
    get { return accountNumber; }
    set { accountNumber = value; }
  }

  // LastName property
  public string LastName
  {
    get { return lastName; }
    set { lastName = value.Length > 15 ? value[..15] : value; }
  }

  // FirstName property
  public string FirstName
  {
    get { return firstName; }
    set { firstName = value.Length > 10 ? value[..10] : value; }
  }

  // Balance property
  public double Balance
  {
    get { return balance; }
    set { balance = value; }
  }
}
